AppCenterCrashes/AppCenterCrashes/MSCrashes.mm

```diff
@@ -276,6 +276,12 @@ + (void)setDelegate:(_Nullable id<MSCrashesDelegate>)delegate {
 
 - (instancetype)init {
   if ((self = [super init])) {
+    NSDictionary *migratedKeys = @{
+      @"kMSCrashesIsEnabledKey" : @"MSACCrashesIsEnabledKey",
+      @"MSAppDidReceiveMemoryWarning" : @"MSACAppDidReceiveMemoryWarning",
+      @"MSUserConfirmation" : @"MSACUserConfirmation"
+    };
+    [MS_USER_DEFAULTS migrateKeys:migratedKeys forService:kMSServiceName];
```

Comment:
🤔 I am thinking `migrateKeys:forService:` is something the service abstraction can do for all services in its init. If services for instance override a `keysToMigrate` method returning the keys to migrate. Actually another reason to do it there is because we are already getting the isEnabled state for the service from the user defaults in the base class and the migration needs to happen before that.

Reply:
Could be a bit convoluted. If we don't do it this way we still need the migration to happen before accessing the `isEnabled` value.

Reply:
I tried this approach before but got stuck. 
Basically it comes down to this: 
1. We don't know what services we have.
2. It means we should use a delegate approach and this delegate has this `getKeysToMigrateMethod`.
3. Assignment of all delegates should happen before `MSUserDefaults` init (at init we migrate).
4. I added all the delegates at module init but that didn't work: some of the modules are initialised earlier than others in our Sasquatch app for example. 
5. Distribute is called -> it sets the delegate -> it then uses MSUserDefaults in the init which triggers the migration of only one module.

Reply:
Probably we could enumerate by modules array somehow..

Reply:
The base class migration / child class key definition would work now but is not ideal simply because we will always have to make sure keys are used only after the migration as you said. We had to deal with such issues when we first introduced the swizzling. We had to declare methods and classes to swizzle depending on services used. We might be able to reuse the same mechanism.

Reply:
Here is how it works. We can have the objective-c runtime do the hard work for us thanks to the [load](https://developer.apple.com/documentation/objectivec/nsobject/1418815-load?language=objc) class method. It's called when classes are loaded by the objective-c runtime. It's synchronous and a class is "loaded" before its categories. Since this is happening before the wrapper initialization no App Center user defaults is get/set and the migration happens after all keys are collected. 

So the wrapper will implement its load method and initialize a class level property exposing a mutable dictionary which will receive the keys to migrate. All services would create a category over the user defaults wrapper. These categories will implement their load method. These load methods will add the keys to be migrated to the user default wrapper dictionary property. Then, when the wrapper will initialize himself it will refer to the dictionary property to migrate all the keys at once.  This should be quite safe.

Now there is a little trick with the categories in our code base. Since we don't require the `-ObjC` flag the categories are not wired automatically at compile time. The only thing you have to do to force it is to implement the categories in classes that we know will be loaded for instance the services main class implementations.

[Example of a category implementing the load method. ](https://github.com/microsoft/appcenter-sdk-apple/blob/d3eb6c942c39be49a6363f28c86f799eb5c1f5c0/AppCenterDistribute/AppCenterDistribute/Internals/MSDistributeAppDelegate.m#L40)
DM me if you need help :).

Reply:
Talked in DM. @annakocheshkova good point, the categories might be useless, just implementing the services `load` methods should be enough. Although I don't think the ordering of class loading between classes of the same lib is guaranteed calling the wrapper in the service load method should force load it.